torch/nn/functional.py

```diff
@@ -5214,6 +5214,7 @@ def multi_head_attention_forward(
     # compute in-projection
     #
     if not use_separate_proj_weight:
+        assert in_proj_weight is not None, "use_separate_proj_weight is False but in_proj_weight is None"
```

Comment:
nit: it's preferred to raise a proper `RuntimeError` here instead of asserting. Something like this:
```python
if not use_separate_proj_weight:
    if in_proj_weight is None:
        raise RuntimeError("Expected in_proj_weight to be set when use_separate_proj_weight=False, but found None")
    q, k, v = _in_projection_packed(query, key, value, in_proj_weight, in_proj_bias)
...
```

I believe this should still pass the mypy checks.

Reply:
Looks like there is quite a few assert in `multi_head_attention_forward`. Are they all supposed to be replaced with RuntimeError?

Reply:
Ah, sure enough there are several. No worries then, we can keep this as-is